src/stdatamodels/validate.py

```diff
@@ -35,6 +39,100 @@ def value_change(path, value, schema, ctx):
     return update
 
 
+def _validate_datatype(validator, schema_datatype, instance, schema):
+    """
+    This extends the ASDF datatype validator to support ndim
+    and max_ndim within individual fields of structured arrays,
+    and handle the absence of the shape field correctly.
+
+    Additionally, dtypes are required to be equivalent, instead
+    of just "safe" to cast.
+    """
+    if isinstance(instance, list):
+        array = ndarray.inline_data_asarray(instance)
+        instance_datatype, _ = ndarray.numpy_dtype_to_asdf_datatype(array.dtype)
+    elif isinstance(instance, dict):
+        if 'datatype' in instance:
+            instance_datatype = instance['datatype']
+        elif 'data' in instance:
+            array = ndarray.inline_data_asarray(instance['data'])
+            instance_datatype, _ = ndarray.numpy_dtype_to_asdf_datatype(array.dtype)
+        else:
+            raise ValidationError("Not an array")
+    elif isinstance(instance, (np.ndarray, ndarray.NDArrayType)):
+        instance_datatype, _ = ndarray.numpy_dtype_to_asdf_datatype(instance.dtype)
+    else:
+        raise ValidationError("Not an array")
+
```

Comment:
Here you are doing `raise ValidationError()` but elsewhere below you are `yield ValidationError()`.  Any reason for this?  Shouldn't they all be `raise`?

Reply:
Oops, good question, these should be yields.  I believe the yield allows the validator to descend into other paths through the combiners and still succeed, while the raise stops it in its tracks.